src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c

```diff
@@ -80,3 +63,238 @@ int32_t GlobalizationNative_IanaIdToWindowsId(const UChar* ianaId, UChar* window
+    //        than are currently in effect.
+    //
+    //        See https://github.com/unicode-org/cldr/blob/master/common/supplemental/metaZones.xml
+    //
+    //        Example:  As of writing this, Africa/Algiers is in the Europe_Central metazone,
+    //                  which has a standard-time name of "Central European Standard Time" (in English).
+    //                  However, in some previous dates, it used the Europe_Western metazone,
+    //                  having the standard-time name of "Western European Standard Time" (in English).
+    //                  Only the *current* name will be returned.
+    //
+    //        TODO: Add a parameter for the timestamp that is used when getting the display names instead of
+    //              getting "now" on the following line.  Everything else should be using this timestamp.
+    //              For now, since TimeZoneInfo presently uses only a single set of display names, we will
+    //              use the names associated with the *current* date and time.
+
+    UDate timestamp = ucal_getNow();
+
+    switch (type)
+    {
+        case TimeZoneDisplayName_Standard:
+            GetTimeZoneDisplayName_FromCalendar(locale, timeZoneId, timestamp, UCAL_STANDARD, result, resultLength, &err);
+            break;
+
+        case TimeZoneDisplayName_DaylightSavings:
+            GetTimeZoneDisplayName_FromCalendar(locale, timeZoneId, timestamp, UCAL_DST, result, resultLength, &err);
+            break;
+
+        case TimeZoneDisplayName_Generic:
+            GetTimeZoneDisplayName_FromPattern(locale, timeZoneId, timestamp, GENERIC_PATTERN_UCHAR, result, resultLength, &err);
+            if (U_SUCCESS(err))
+            {
+                FixupTimeZoneGenericDisplayName(locale, timeZoneId, timestamp, result, &err);
+            }
+            break;
+
+        case TimeZoneDisplayName_GenericLocation:
+            GetTimeZoneDisplayName_FromPattern(locale, timeZoneId, timestamp, GENERIC_LOCATION_PATTERN_UCHAR, result, resultLength, &err);
+            break;
+
+        case TimeZoneDisplayName_ExemplarCity:
+            GetTimeZoneDisplayName_FromPattern(locale, timeZoneId, timestamp, EXEMPLAR_CITY_PATTERN_UCHAR, result, resultLength, &err);
+            break;
+
+        default:
+            return UnknownError;
+    }
+
+    return GetResultCode(err);
+}
+
+/*
+Private function to get the standard and daylight names from the ICU Calendar API.
+*/
+void GetTimeZoneDisplayName_FromCalendar(const char* locale, const UChar* timeZoneId, const UDate timestamp, UCalendarDisplayNameType type, UChar* result, int32_t resultLength, UErrorCode* err)
+{
+    // Examples: "Pacific Standard Time"  (standard)
+    //           "Pacific Daylight Time"  (daylight)
+
+    // (-1 == timeZoneId is null terminated)
+    UCalendar* calendar = ucal_open(timeZoneId, -1, locale, UCAL_DEFAULT, err);
+    if (U_FAILURE(*err)) goto exit;
```

Comment:
>exit [](start = 30, length = 4)

this will jump to exit and close the calendar while the open operation is failed. Alos, I don't think we need to use go to in such short method. you may implment that as the following:

UCalendar* calendar = ucal_open(timeZoneId, -1, locale, UCAL_DEFAULT, err);
if (U_SUCCESS(*err))
{
    ucal_setMillis(calendar, timestamp, err);
    if (U_SUCCESS(*err))
    {
        ucal_getTimeZoneDisplayName(calendar, type, locale, result, resultLength, err);
    }
    ucal_close(calendar);
}

Reply:
I had used ICU's tests and examples to figure out if one should *always* call `ucal_close`, or if one should only close it if the status indicated success.  Most show it as always closing, where a few examples return early without closing, so they have it both ways.

Digging a bit deeper into the implementation in ICU sources, it seems `ucal_close` is guarded - it only calls `delete` on a calendar instance if it's not a `nullptr`, so it's safe to call in all cases.  But also, the `ucal_open` function only allocates the calendar instance when it also returns success.  Thus, it doesn't matter either way.

Interestingly, `udat_close` isn't guarded for `nullptr` - but [from what I can assertain](https://stackoverflow.com/a/6731484/634824) delete on a nullptr has no effect anyway.

I'll still adjust as you suggest.  Its a few less instructions at any rate.

Reply:
The issue is judging current ICU implementation is not enough because things can change in the future. We should be writing the code in the way that shouldn't fail in the future too. it is best practice to only close handles which is valid. 

> I'll still adjust as you suggest. Its a few less instructions at any rate.

Thanks!